File: sdks/typescript/src/transport/ws.ts
import {
  type Envelope,
  decodeEnvelope,
  encodeEnvelope,
  MsgType,
} from "../protocol/codec.js";

type WebSocketLike = {
  binaryType: string;
  readyState: number;
  close(): void;
  send(data: ArrayBuffer | Uint8Array): void;
  addEventListener(type: string, listener: (event: any) => void): void;
  removeEventListener(type: string, listener: (event: any) => void): void;
};

function createWebSocket(url: string): WebSocketLike {
  // Use native WebSocket in browser, ws in Node.js
  if (typeof globalThis.WebSocket !== "undefined") {
    return new globalThis.WebSocket(url) as unknown as WebSocketLike;
  }
  // Dynamic require for Node.js ws package
  // eslint-disable-next-line @typescript-eslint/no-require-imports
  const WS = require("ws");
  return new WS(url) as WebSocketLike;
}

export type MessageHandler = (envelope: Envelope) => void;

export class WsTransport {
  private ws: WebSocketLike | null = null;
  private url: string;
  private token: string;
  private handlers = new Map<number, MessageHandler[]>(); // type -> handlers
  private pendingRequests = new Map<
    number,
    { resolve: (env: Envelope) => void; reject: (err: Error) => void }
  >();
  private nextId = 1;
  private reconnectAttempts = 0;
  private maxReconnectAttempts = 5;
  private connected = false;
  private connectPromise: Promise<void> | null = null;
  private intentionallyClosed = false;

  constructor(url: string, token: string) {
    this.url = url;
    this.token = token;
  }

  async connect(): Promise<void> {
    if (this.connected) return;
    if (this.connectPromise) return this.connectPromise;

    this.connectPromise = new Promise<void>((resolve, reject) => {
      const wsUrl = `${this.url}?token=${encodeURIComponent(this.token)}`;
      this.ws = createWebSocket(wsUrl);
      this.ws.binaryType = "arraybuffer";

      this.ws.addEventListener("open", () => {
        this.connected = true;
        this.reconnectAttempts = 0;
        this.connectPromise = null;
        resolve();
      });

      this.ws.addEventListener("message", (event: any) => {
        const raw = event.data ?? event;
        let bytes: Uint8Array;
        if (raw instanceof ArrayBuffer) {
          bytes = new Uint8Array(raw);
        } else if (raw instanceof Uint8Array) {
          bytes = raw;
        } else {
          // Node.js Buffer
          bytes = new Uint8Array(raw.buffer, raw.byteOffset, raw.byteLength);
        }
        this.handleMessage(bytes);
      });

      this.ws.addEventListener("close", () => {
        this.connected = false;
        this.connectPromise = null;
        this.attemptReconnect();
      });

      this.ws.addEventListener("error", (err: any) => {
        this.connectPromise = null;
        if (!this.connected) {
          reject(err);
        }
      });
    });

    return this.connectPromise;
  }

  close(): void {
    this.intentionallyClosed = true;
    if (this.ws) {
      this.ws.close();
      this.ws = null;
    }
    this.connected = false;

    // Reject all pending requests
    for (const [, pending] of this.pendingRequests) {
      pending.reject(new Error("Connection closed"));
    }
    this.pendingRequests.clear();
  }

  send(envelope: Envelope): void {
    if (!this.ws || !this.connected) {
      throw new Error("WebSocket not connected. Ensure the sandbox is running and call connect() first.");
    }
    const data = encodeEnvelope(envelope);
    this.ws.send(data);
  }

  async request(
    type: number,
    channel: number,
    payload: Uint8Array,
    timeout = 30_000
  ): Promise<Envelope> {
    const id = this.nextId++;

    return new Promise<Envelope>((resolve, reject) => {
      const timer = setTimeout(() => {
        this.pendingRequests.delete(id);
        reject(new Error(`WebSocket request timeout after ${timeout}ms (id=${id}). The sandbox sidecar may be unresponsive.`));
      }, timeout);

      this.pendingRequests.set(id, {
        resolve: (env) => {
          clearTimeout(timer);
          resolve(env);
        },
        reject: (err) => {
          clearTimeout(timer);
          reject(err);
        },
      });

      this.send({ type, channel, id, payload });
    });
  }

  on(type: number, handler: MessageHandler): () => void {
    const existing = this.handlers.get(type) ?? [];
    existing.push(handler);
    this.handlers.set(type, existing);

    return () => {
      const handlers = this.handlers.get(type);
      if (handlers) {
        const idx = handlers.indexOf(handler);
        if (idx !== -1) handlers.splice(idx, 1);
      }
    };
  }

  private handleMessage(data: Uint8Array): void {
    let envelope: Envelope;
    try {
      envelope = decodeEnvelope(data);
    } catch {
      return;
    }

    // Handle ping/pong
    if (envelope.type === MsgType.Ping) {
      this.send({
        type: MsgType.Pong,
        channel: 0,
        id: envelope.id,
        payload: new Uint8Array(0),
      });
      return;
    }

    // Check if this is a response to a pending request
    if (envelope.id > 0) {
      const pending = this.pendingRequests.get(envelope.id);
      if (pending) {
        this.pendingRequests.delete(envelope.id);
        if (envelope.type === MsgType.Error) {
          pending.reject(new Error(`Server error`));
        } else {
          pending.resolve(envelope);
        }
        return;
      }
    }

    // Dispatch to type handlers
    const handlers = this.handlers.get(envelope.type);
    if (handlers) {
      for (const handler of handlers) {
        handler(envelope);
      }
    }
  }

  private attemptReconnect(): void {
    if (this.intentionallyClosed) return;
    if (this.reconnectAttempts >= this.maxReconnectAttempts) return;

    this.reconnectAttempts++;
    const delay = Math.min(1000 * Math.pow(2, this.reconnectAttempts - 1), 30_000);

    setTimeout(() => {
      if (this.intentionallyClosed || this.connected) return;
      this.connect().catch(() => {
        // connect() failure will trigger another close event, which retries
      });
    }, delay);
  }
}

File: sdks/python/src/xgen_sandbox/transport/ws.py
from __future__ import annotations

import asyncio
from typing import Callable
from urllib.parse import urlencode

import websockets
import websockets.asyncio.client

from ..protocol.codec import (
    MsgType,
    Envelope,
    encode_envelope,
    decode_envelope,
)

MessageHandler = Callable[[Envelope], None]


class WsTransport:
    def __init__(self, url: str, token: str, max_reconnect_attempts: int = 5) -> None:
        self._url = url
        self._token = token
        self._ws: websockets.asyncio.client.ClientConnection | None = None
        self._handlers: dict[int, list[MessageHandler]] = {}
        self._pending: dict[int, asyncio.Future[Envelope]] = {}
        self._next_id = 1
        self._connected = False
        self._recv_task: asyncio.Task | None = None
        self._intentionally_closed = False
        self._reconnect_attempts = 0
        self._max_reconnect_attempts = max_reconnect_attempts

    async def connect(self) -> None:
        if self._connected:
            return
        ws_url = f"{self._url}?{urlencode({'token': self._token})}"
        self._ws = await websockets.asyncio.client.connect(ws_url)
        self._connected = True
        self._reconnect_attempts = 0
        self._intentionally_closed = False
        self._recv_task = asyncio.get_event_loop().create_task(self._recv_loop())

    def close(self) -> None:
        self._intentionally_closed = True
        self._connected = False
        if self._recv_task and not self._recv_task.done():
            self._recv_task.cancel()
        if self._ws:
            asyncio.get_event_loop().create_task(self._ws.close())
            self._ws = None
        # Reject pending requests
        for fut in self._pending.values():
            if not fut.done():
                fut.set_exception(ConnectionError("Connection closed"))
        self._pending.clear()

    def send(self, envelope: Envelope) -> None:
        if not self._ws or not self._connected:
            raise ConnectionError("Not connected")
        data = encode_envelope(envelope)
        asyncio.get_event_loop().create_task(self._ws.send(data))

    async def send_async(self, envelope: Envelope) -> None:
        if not self._ws or not self._connected:
            raise ConnectionError("Not connected")
        data = encode_envelope(envelope)
        await self._ws.send(data)

    async def request(
        self,
        msg_type: int,
        channel: int,
        payload: bytes,
        timeout: float = 30.0,
    ) -> Envelope:
        msg_id = self._next_id
        self._next_id += 1

        loop = asyncio.get_event_loop()
        fut: asyncio.Future[Envelope] = loop.create_future()
        self._pending[msg_id] = fut

        await self.send_async(
            Envelope(type=msg_type, channel=channel, id=msg_id, payload=payload)
        )

        try:
            return await asyncio.wait_for(fut, timeout=timeout)
        except asyncio.TimeoutError:
            self._pending.pop(msg_id, None)
            raise TimeoutError(f"Request timeout (id={msg_id})")

    def on(self, msg_type: int, handler: MessageHandler) -> Callable[[], None]:
        handlers = self._handlers.setdefault(msg_type, [])
        handlers.append(handler)

        def cleanup() -> None:
            try:
                handlers.remove(handler)
            except ValueError:
                pass

        return cleanup

    # -- internal --

    async def _recv_loop(self) -> None:
        assert self._ws is not None
        try:
            async for message in self._ws:
                if isinstance(message, str):
                    continue
                self._handle_message(message)
        except websockets.exceptions.ConnectionClosed:
            pass
        except asyncio.CancelledError:
            return  # intentional cancellation, don't reconnect
        finally:
            self._connected = False
            asyncio.ensure_future(self._attempt_reconnect())

    def _handle_message(self, data: bytes) -> None:
        try:
            envelope = decode_envelope(data)
        except Exception:
            return

        # Respond to pings
        if envelope.type == MsgType.Ping:
            self.send(
                Envelope(type=MsgType.Pong, channel=0, id=envelope.id, payload=b"")
            )
            return

        # Check pending requests by id
        if envelope.id > 0:
            fut = self._pending.pop(envelope.id, None)
            if fut is not None and not fut.done():
                if envelope.type == MsgType.Error:
                    fut.set_exception(RuntimeError("Server error"))
                else:
                    fut.set_result(envelope)
                return

        # Dispatch to type handlers
        handlers = self._handlers.get(envelope.type)
        if handlers:
            for handler in list(handlers):
                handler(envelope)

    async def _attempt_reconnect(self) -> None:
        if self._intentionally_closed:
            return
        if self._reconnect_attempts >= self._max_reconnect_attempts:
            return

        self._reconnect_attempts += 1
        delay = min(2 ** (self._reconnect_attempts - 1), 30)
        await asyncio.sleep(delay)

        if self._intentionally_closed or self._connected:
            return

        try:
            await self.connect()
        except Exception:
            pass  # connect failure triggers recv_loop exit which retries
